x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java

```diff
@@ -838,13 +921,66 @@ private void innerRefresh(String tokenDocId, Authentication userAuth, ActionList
+                                                                                           Authentication userAuth) {
+        final Map<String, Object> refreshTokenSrc = (Map<String, Object>) source.get("refresh_token");
+        final Map<String, Object> userTokenSource = (Map<String, Object>)
+            ((Map<String, Object>) source.get("access_token")).get("user_token");
+        final Integer version = (Integer) userTokenSource.get("version");
+        Version authVersion = Version.fromId(version);
+        final Boolean refreshed = (Boolean) refreshTokenSrc.get("refreshed");
+        if (refreshed) {
+            if (authVersion.onOrAfter(Version.V_8_0_0)) {
+                final Long refreshedEpochMilli = (Long) refreshTokenSrc.get("refresh_time");
+                final Instant refreshTime = refreshedEpochMilli == null ? null : Instant.ofEpochMilli(refreshedEpochMilli);
+                final String supersededBy = (String) refreshTokenSrc.get("superseded_by");
+                if (supersededBy == null) {
+                    return Optional.of(invalidGrantException("token document is missing superseded by value"));
+                } else if (refreshTime == null) {
+                    return Optional.of(invalidGrantException("token document is missing refresh time value"));
+                } else if (clock.instant().isAfter(refreshTime.plus(4L, ChronoUnit.SECONDS))) {
+                    return Optional.of(invalidGrantException("token has already been refreshed"));
+                }
+            } else {
+                return Optional.of(invalidGrantException("token has already been refreshed"));
+            }
+        }
+        return checkClient(refreshTokenSrc, userAuth);
+    }
+
+    /**
+     * Checks if a refreshed token is eligible to be refreshed again. This is only allowed for versions after 7.0.0 and
+     * when the refresh_token contains the refresh_time and superseded_by fields and it has been refreshed in the
+     * previous 4 seconds
+     */
+    private boolean eligibleForMultiRefresh(Map<String, Object> source) {
+        final Map<String, Object> refreshTokenSrc = (Map<String, Object>) source.get("refresh_token");
+        final Map<String, Object> userTokenSource = (Map<String, Object>)
+            ((Map<String, Object>) source.get("access_token")).get("user_token");
+        final Integer version = (Integer) userTokenSource.get("version");
+        Version authVersion = Version.fromId(version);
+        final Long refreshedEpochMilli = (Long) refreshTokenSrc.get("refresh_time");
+        final Instant refreshTime = refreshedEpochMilli == null ? null : Instant.ofEpochMilli(refreshedEpochMilli);
+        final String supersededBy = (String) refreshTokenSrc.get("superseded_by");
+        return authVersion.onOrAfter(Version.V_8_0_0)
+            && supersededBy != null
+            && refreshTime != null
+            && clock.instant().isAfter(refreshTime.plus(4L, ChronoUnit.SECONDS)) == false;
```

Comment:
I think we need a different instant here to validate against rather than the current one